src/agent/src/device/mod.rs

```diff
@@ -0,0 +1,1109 @@
+        manager
+    };
+}
+
+#[instrument]
+pub async fn add_devices(
+    logger: &Logger,
+    devices: &[Device],
+    spec: &mut Spec,
+    sandbox: &Arc<Mutex<Sandbox>>,
+) -> Result<()> {
+    let mut dev_updates = HashMap::<&str, DevUpdate>::with_capacity(devices.len());
+
+    for device in devices.iter() {
+        validate_device(logger, device, sandbox).await?;
+        if let Some(handler) = DEVICE_HANDLERS.handler(&device.type_) {
+            let mut ctx = DeviceContext { logger, sandbox };
+
+            match handler.device_handler(device, &mut ctx).await {
+                Ok(update) => {
+                    if let Some(dev_update) = update.dev {
+                        if dev_updates
+                            .insert(&device.container_path, dev_update.clone())
+                            .is_some()
+                        {
+                            return Err(anyhow!(
+                                "Conflicting device updates for {}",
+                                &device.container_path
+                            ));
+                        }
+
+                        // Update cgroup to allow all devices added to guest.
+                        insert_devices_cgroup_rule(logger, spec, &dev_update.info, true, "rwm")
+                            .context("Update device cgroup")?;
+                    }
+
+                    let mut sb = sandbox.lock().await;
+                    for (host, guest) in update.pci {
+                        if let Some(other_guest) = sb.pcimap.insert(host, guest) {
+                            return Err(anyhow!(
+                                "Conflicting guest address for host device {} ({} versus {})",
+                                host,
+                                guest,
+                                other_guest
+                            ));
+                        }
+                    }
+                }
+                Err(e) => {
+                    error!(logger, "failed to add devices, error: {e:?}");
+                    return Err(e);
+                }
+            }
+        } else {
+            return Err(anyhow!(
+                "Failed to find the device handler {}",
+                device.type_
+            ));
+        }
+    }
+
+    if let Some(process) = spec.process_mut() {
+        let env_vec: &mut Vec<String> =
+            &mut process.env_mut().get_or_insert_with(Vec::new).to_vec();
+        update_env_pci(env_vec, &sandbox.lock().await.pcimap)?
+    }
+    update_spec_devices(logger, spec, dev_updates)
+}
+
+#[instrument]
+async fn validate_device(
+    logger: &Logger,
+    device: &Device,
+    sandbox: &Arc<Mutex<Sandbox>>,
+) -> Result<()> {
+    // log before validation to help with debugging gRPC protocol version differences.
+    info!(
+        logger,
+        "device-vm-path: {}, device-container-path: {}, device-options: {:?}",
```

Comment:
I think we should keep the two helpful fields: "device-id" and "device-type" when checking the log message.

Reply:
Done.